certbot-nginx/certbot_nginx/configurator.py

```diff
@@ -621,21 +633,22 @@ def _enable_redirect(self, domain, unused_options):
         if vhost is None:
             logger.info("No matching insecure server blocks listening on port %s found.",
                 self.DEFAULT_LISTEN_PORT)
+            return
+
+        if self._has_certbot_redirect(vhost):
+            logger.info("Traffic on port %s already redirecting to ssl in %s",
+                self.DEFAULT_LISTEN_PORT, vhost.filep)
+        elif vhost.has_redirect():
+            if not self._has_certbot_redirect_comment(vhost):
+                self._add_redirect_block(vhost, active=False, use_if=(vhost.ssl_copy is None))
```

Comment:
Given the current implementation and possible ways of reaching this code, this logic is correct. I think it makes sense to change it in the future if and when either or those things change, for example while adding an enhance subcommand. I generally prefer to avoid parsing-based code and rely instead of program-flow and passing around variables that we have set and are sure of.

Reply:
> I generally prefer to avoid parsing-based code and rely instead of program-flow and passing around variables that we have set and are sure of.

I wish we could get our parsing to the point where we were confident in that as well. While very similar to the case I described above, this code also unnecessarily uses a conditional redirect in the case the user already had separate HTTP and HTTPS blocks.

Due to this and discussions about adding an `enhance` subcommand sometime in the near future, I think this should be fixed although I'm fine with not doing it in this PR. If we wait, we should make an issue to properly track this.